client/src/app/worker/autoupdate-stream.ts

```diff
@@ -0,0 +1,259 @@
+    ) {}
+
+    /**
+     * Clones this stream with the specified subscriptions
+     *
+     * @param subscriptions The subscriptions handled by the created stream
+     */
+    public cloneWithSubscriptions(subscriptions: AutoupdateSubscription[]): AutoupdateStream {
+        return new AutoupdateStream(subscriptions, this.queryParams, this.endpoint);
+    }
+
+    /**
+     * Closes the stream
+     */
+    public abort() {
+        if (this.abortCtrl !== undefined) {
+            // @ts-ignore
+            this.abortCtrl.abort();
+        }
+    }
+
+    /**
+     * Opens a new connection to autoupdate.
+     * Also this function registers this stream inside all subscriptions
+     * handled by this stream.
+     *
+     * resolves when fetch connection is closed
+     */
+    public async start(
+        force?: boolean
+    ): Promise<{ stopReason: 'error' | 'aborted' | 'unused' | 'resolved' | 'in-use'; error?: any }> {
+        if (this.active && !force) {
+            return { stopReason: `in-use` };
+        } else if (this.active && force) {
+            this.abort();
+        }
+
+        this.restart = false;
+        this.error = null;
+        try {
+            await this.doRequest();
+            this.active = false;
+        } catch (e) {
+            this.active = false;
+            if (e.name !== `AbortError`) {
+                console.error(e);
+
+                return { stopReason: `error`, error: this.error };
+            } else if (this.restart) {
+                return await this.start();
+            }
+
+            return { stopReason: this.activeSubscriptions?.length ? `aborted` : `unused`, error: this.error };
+        }
+
+        if (this.error) {
+            return { stopReason: `error`, error: this.error };
+        }
+
+        return { stopReason: `resolved` };
+    }
+
+    private async doRequest() {
+        this.active = true;
+
+        if (this.activeSubscriptions === null) {
+            this.activeSubscriptions = [];
+            for (let subscription of this.subscriptions) {
+                this.activeSubscriptions.push(subscription);
+                subscription.stream = this;
+            }
+        }
+
+        const headers: any = {
+            'Content-Type': `application/json`
+        };
+
+        if (this.endpoint.authToken) {
+            headers.authentication = this.endpoint.authToken;
+        }
+
+        this.abortCtrl = new AbortController();
+
+        const response = await fetch(this.endpoint.url + this.queryParams, {
+            signal: this.abortCtrl.signal,
+            method: this.endpoint.method,
+            headers,
+            body: JSON.stringify(this.subscriptions.map(s => s.request))
+        });
+
+        const reader = response.body.getReader();
+        let next = null;
+        let result: ReadableStreamDefaultReadResult<Uint8Array>;
+        while (!(result = await reader.read()).done) {
```

Comment:
This is too long and too nested. Maybe create helper functions if it makes sense